### PR TITLE
Fix unknown c++ flag detection in CMake

### PR DESCRIPTION
**Pull request by cyyever**
Unknown -Wno-XXX flags are still appended to GCC via append_cxx_flag_if_supported  because of the behavior mentioned in GCC document:
```
When an unrecognized warning option is requested (e.g., -Wunknown-warning), 
GCC emits a diagnostic stating that the option is not recognized. 
However, if the -Wno- form is used, the behavior is slightly different: 
no diagnostic is produced for -Wno-unknown-warning unless other diagnostics are being produced. 
This allows the use of new -Wno- options with old compilers, 
but if something goes wrong, the compiler warns that an unrecognized option is present. 
``` 
This PR tries to fix by detection the flag of the -WXXX form. Unfortunately, third_party/fbgemm/CMakeLists.txt redefines append_cxx_flag_if_supported and our version is overwritten. As a result, we have to re-include utils.cmake to overwrite it again.
 

cc @malfet  @ezyang

### CODE REVIEW

File: cmake/public/utils.cmake [hunk 1]
```diff
@@ -537,11 +537,16 @@ endfunction()
...
 # Also define HAS_UPPER_CASE_FLAG_NAME variable
 # Usage:
-#   append_cxx_flag_if_supported("-Werror" CMAKE_CXX_FLAGS)
-function(append_cxx_flag_if_supported flag outputvar)
+#   append_torch_cxx_flag_if_supported("-Werror" CMAKE_CXX_FLAGS)
+function(append_torch_cxx_flag_if_supported flag outputvar)
     string(TOUPPER "HAS${flag}" _FLAG_NAME)
     string(REGEX REPLACE "[=-]" "_" _FLAG_NAME "${_FLAG_NAME}")
-    check_cxx_compiler_flag("${flag}" ${_FLAG_NAME})
+    if(NOT CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
```
Comment (malfet) [inline cmake/public/utils.cmake:544]: are you sure that clang behaves the same way?
Also, please add a comment explaining why this is necessary 
```suggestion
    if(CMAKE_CXX_COMPILER_ID STREQUAL "GCC")
```

**Review by malfet — APPROVED** [review]
Looks ok, but instead of renaming the function, do you mind submitting similar PR against fbgemm? And in this pr update fbgemm to new hash

File: cmake/public/utils.cmake (hunk 1 above)
Comment (cyyever) [reply cmake/public/utils.cmake]: After a test I found it is GCC specific behavior.

**Comment by cyyever** [comment]
> Looks ok, but instead of renaming the function, do you mind submitting similar PR against fbgemm? And in this pr update fbgemm to new hash

I want to de-couple the dependency to fbgemm in this function. It took me half a hour to find out why  the local modifications and it is better to avoid future confusions from others,

**Pull request closed without merging**